gplugins/lumerical/eme.py

```diff
@@ -0,0 +1,1063 @@
+        Updates both Lumerical session and simulation settings.
+
+        Parameters:
+            plot: Plot and save convergence results
+
+        Returns:
+            Convergence results
+        """
+        s = self.session
+        cs = self.convergence_settings
+        ss = self.simulation_settings
+
+        s21 = []
+        s11 = []
+        num_cells = []
+        converged = False
+        while not converged:
+            s.switchtolayout()
+            s.setnamed("EME", "cells", np.array(ss.group_cells))
+            s.run()
+            s.emepropagate()
+            S = s.getresult("EME", "user s matrix")
+            s11.append(abs(S[0, 0]) ** 2)
+            s21.append(abs(S[1, 0]) ** 2)
+            num_cells.append(ss.group_cells[1:-1])
+
+            for i in range(1, len(ss.group_cells) - 1):
+                ss.group_cells[i] += 1
+
+            # Check whether convergence has been reached
+            if len(s21) > cs.passes or len(s11) > cs.passes:
+                # Calculate maximum diff in sparams
+                sparam_diff = max(
+                    [
+                        max(abs(s21[-1] - np.array(s21[-(cs.passes + 1) : -1]))),
+                        max(abs(s11[-1] - np.array(s11[-(cs.passes + 1) : -1]))),
+                    ]
+                )
+                if plot:
+                    logger.info(
+                        f"Num cells: {ss.group_cells} | S-Param Diff: {sparam_diff}"
+                    )
+                if sparam_diff < cs.sparam_diff:
+                    converged = True
+                else:
+                    converged = False
+
+        num_cells = np.array(num_cells)
+        if plot:
+            plt.figure()
+            plt.plot(list(num_cells[:, 0]), s21)
+            plt.plot(list(num_cells[:, 0]), s11)
+            plt.xticks(list(num_cells[:, 0]), [f"{list(row)}" for row in num_cells])
+            plt.setp(plt.xticks()[1], rotation=75, horizontalalignment="center")
+            plt.legend(["|S21|^2", "|S11|^2"])
+            plt.grid("on")
+            plt.xlabel("Number of Cells")
+            plt.ylabel("Magnitude")
+            plt.title(f"Cell Convergence | Wavelength={ss.wavelength}um")
+            plt.tight_layout()
+            plt.savefig(
+                str(
+                    self.dirpath
+                    / f"{self.component.name}_convergence"
+                    / f"{self.component.name}_eme_cell_convergence.png"
+                )
+            )
+
+        convergence_data = pd.DataFrame.from_dict(
+            {"num_cells": list(num_cells[:, 0]), "s21": list(s21), "s11": list(s11)}
+        )
+        convergence_data.to_csv(
+            str(
+                self.dirpath
+                / f"{self.component.name}_convergence"
+                / f"{self.component.name}_eme_cell_convergence.csv"
+            )
+        )
+        return convergence_data
+
+    def update_mode_convergence(self, plot: bool = False) -> pd.DataFrame:
```

Comment:
are these `update_*_convergence` different convergence parameters that are swept? This seems like not the right way to do a multi-axis sweep (especially when there might be interdependencies on convergence). we can discuss